### PR TITLE
Add GetPreferredAllocation() call to the v1beta1 device plugin API

### LINKED ISSUE #83483: TopologyManager: Add support for device-specific topology constraints
The current `TopologyManager` only allows devices to be aligned by NUMA node. However, certain devices have internal constraints which dictate how they should be aligned when multiple devices of the same type are requested.

This issue tracks the effort to add support for such "inter-device" topology constraints when allocating devices to a container. The proposed solution allows each device plugin to specify preferred device allocations based on internal topology constraints, while still leaving the final allocation decision up to the `devicemanager`. 

KEP Update:https://github.com/kubernetes/enhancements/pull/1121
Implementation PR: TBD

### PR DESCRIPTION
**Pull request by klueska**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:
This PR adds support for a new, optional call in the `v1beta1` device plugin API called `GetPreferredAllocation()`. Because it is an optional extension (and not a change) to the API, it does not require a version bump.

Details of this new API call can be found in the following KEP update: https://github.com/kubernetes/enhancements/pull/1121

Please review commit-by-commit.

**Which issue(s) this PR fixes**:

Fixes #83483

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Added a new `GetPreferredAllocation()` call to the `v1beta1` device plugin API.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
[KEP]: https://github.com/kubernetes/enhancements/pull/1121
```

### CODE REVIEW

**Comment by klueska** [comment]
/assign @derekwaynecarr 
To approve the API changes

The only reason this this currently a WIP is because I want to add more extensive testing before merging. We should be able to work on approving the API changes without this though.

**Comment by klueska** [comment]
/assign @mkumatag 
/assign @listx
For approval in small change to `test/images/OWNERS` in support of this new API.

File: pkg/kubelet/cm/devicemanager/manager.go
Comment (nolancon) [inline pkg/kubelet/cm/devicemanager/manager.go:711]: I would have thought allocation from the preferred list would take place before the aligned list. Is this because we can't assume preferred devices are aligned also?

File: pkg/kubelet/cm/devicemanager/manager.go
Comment (nolancon) [inline pkg/kubelet/cm/devicemanager/manager.go:975]: Could you explain why this check and returning an error is necessary here? It's not immediately clear to me, thanks.

**Review by nolancon — COMMENTED** [review]
Just one nit and a couple of questions for my own understanding.

File: pkg/kubelet/cm/devicemanager/manager.go
Comment (klueska) [reply pkg/kubelet/cm/devicemanager/manager.go:711]: Exactly. The goal is to first honour allocations based on topologymanager alignment, and only then defer to the plugin to help decide how to alllocate the rest.

File: pkg/kubelet/cm/devicemanager/manager.go
Comment (klueska) [reply pkg/kubelet/cm/devicemanager/manager.go:975]: I didn't actually look too closely at this when adding it.

It is copy-pasted from the `callPreStartContainerIfNeeded()` function above, which is similar in nature to this one (i.e. only call out to the plugin it if the options set on the plugin indicate to).

Let me take a closer look at everything in this function to see if we need it.

Comment (klueska) [reply pkg/kubelet/cm/devicemanager/manager.go:975]: Just looked. Good catch. Not only is it unnecessary, it would have caused problems once I started writing tests that actually exercise this function (because it's highly likely that no devices will be allocated at the time this is called!).

**Review by vpickard — COMMENTED** [review]
Looks good, few minor comment. Thanks!

File: pkg/kubelet/cm/devicemanager/device_plugin_stub.go
Comment (klueska) [reply pkg/kubelet/cm/devicemanager/device_plugin_stub.go]: Done

Comment (klueska) [reply pkg/kubelet/cm/devicemanager/device_plugin_stub.go]: Done

File: pkg/kubelet/cm/devicemanager/manager.go
Comment (klueska) [reply pkg/kubelet/cm/devicemanager/manager.go]: Updated with detailed comments in latest

File: pkg/kubelet/cm/devicemanager/manager.go
Comment (klueska) [reply pkg/kubelet/cm/devicemanager/manager.go]: Fixed in latest

**Review by nolancon — COMMENTED** [review]
One question. Changes look good otherwise.

File: pkg/kubelet/cm/devicemanager/manager.go
Comment (klueska) [reply pkg/kubelet/cm/devicemanager/manager.go:691]: No. If they are equal then I want to fall through and just allocate all of the needed ones from the aligned set directly.

**Comment by derekwaynecarr** [comment]
nice job, tests look good too.

/approve
/lgtm

**Review by mkumatag — CHANGES_REQUESTED** [review]
Please bump the version here: https://github.com/kubernetes/kubernetes/blob/master/test/images/sample-device-plugin/VERSION

**Comment by klueska** [comment]
> Please bump the version here: https://github.com/kubernetes/kubernetes/blob/master/test/images/sample-device-plugin/VERSION

Thanks for the review. VERSION has been updated here: https://github.com/kubernetes/kubernetes/pull/92665/commits/cbd405d85cca84344f3acc21c86a9eff00ff58b2#diff-efcabe360f36f7a76db00afca0404498

**Comment by mkumatag** [comment]
/lgtm
/approve
/hold

Only for `k/k/test/images` content, adding hold for the remaining

**Pull request merged**
